_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+00000000: 7f45 4c46                                .ELF
+```
+
+Compilers and IDEs will sometimes use different extensions for ELF artifacts. The most common ones
+you will see are `.elf`, `.out`, or `.axf`.
+
+### DWARF
+
+DWARF is a standardized format[^3] for representing debug information (i.e symbol names, struct
+layouts, how to recover backtraces, etc). The format was developed in the late 1980s as
+part of Unix System V Release 4 (SVR4):
+
+> Fun Fact: "The name DWARF is something of a pun, since it was developed along with the ELF object file
+> format. The name is an acronym for “Debugging With Arbitrary Record Formats”".[^2]
+
+When a program is compiled with debug information enabled (e.g. with the `-g` CFLAG when using GCC
+or Clang), this information will be written into special `.debug_` ELF sections that look like:
+
+```
+$ readelf -S <ELF>
+  ...
+  [ 8] .debug_info       PROGBITS        00000000 0200f4 00980b 00      0   0  1
+  [ 9] .debug_abbrev     PROGBITS        00000000 0298ff 001adf 00      0   0  1
+  [10] .debug_loc        PROGBITS        00000000 02b3de 003bf7 00      0   0  1
+  [11] .debug_aranges    PROGBITS        00000000 02efd5 0004f0 00      0   0  1
+  [12] .debug_ranges     PROGBITS        00000000 02f4c5 0007d8 00      0   0  1
+  [13] .debug_macro      PROGBITS        00000000 02fc9d 0031dd 00      0   0  1
+  [14] .debug_line       PROGBITS        00000000 032e7a 005447 00      0   0  1
+  [15] .debug_str        PROGBITS        00000000 0382c1 00dedd 01  MS  0   0  1
+```
+
+### stabs, COFF, XCOFF
+
+While rather infrequent, if you are working with custom compilers (such as
+those shipped with some DSPs), you may see binary / debug information emitted in alternative formats such as
+stabs, COFF, or XCOFF. These are all legacy formats that have largely been superseded by ELF +
+DWARF. I view it as a red-flag in the evaluation process if a vendors compiler emits information in
+these formats. At this point, a lot of binary inspection tools have deprecated support
+for these formats. This makes it harder to debug and workaround issues on chips using these formats.
+
+## Examples
+
+All the example code used in the sections below can be found in it's entirety
+[on Github](https://github.com/memfault/interrupt/tree/master/example/best-of-binutils). If you
+would like to follow along and build any of the sources yourself you can find more information there.
+
+## GNU Binutils
+
+GNU Binutils is a collection of supporting tools that work alongside GNU GCC to inspect and edit binaries.
+We'll go into details below but most of the tools operate on ELFs. On linux and for embedded
+devices both individual object files (`.o`) and the final linked output (the ELF) are in this format.
+
+### Compiling Binutils From Source
+
+Sometimes it can be fun (or useful) to compile the latest GNU Binutil to get a preview of
+unreleased features or customize the supported architecture targets that are supported:
+
+```
```

Comment:
If you want to also try compiling from source, you can try the below steps. 

....

If you'd rather use a pre-compiled version, you can run either `brew install binutils` or `apt install binutils`

Reply:
neat, looks like its a multi-arch target too ... didnt realize